sdk/src/main/scala/com/horizen/validation/ConsensusValidator.scala

```diff
@@ -0,0 +1,82 @@
+package com.horizen.validation
+import java.time.Instant
+
+import com.horizen.SidechainHistory
+import com.horizen.block.SidechainBlock
+import com.horizen.chain.SidechainBlockInfo
+import com.horizen.consensus.{NonceConsensusEpochInfo, _}
+import scorex.core.block.Block
+import scorex.util.ScorexLogging
+
+import scala.util.Try
+
+class ConsensusValidator extends HistoryBlockValidator with ScorexLogging {
+  override def validate(block: SidechainBlock, history: SidechainHistory): Try[Unit] = Try {
+    if (history.isGenesisBlock(block.id)) {
+      validateGenesisBlock(block, history)
+    }
+    else {
+      validateNonGenesisBlock(block, history)
+    }
+  }
+
+  private def validateGenesisBlock(block: SidechainBlock, history: SidechainHistory): Unit = {
+    if (block.timestamp != history.params.sidechainGenesisBlockTimestamp) {
+      throw new IllegalArgumentException(s"Genesis block timestamp ${block.timestamp} is differ than expected timestamp from configuration ${history.params.sidechainGenesisBlockTimestamp}")
+    }
+
+    val vrfSignIsNotCorrect = false //shall be implemented as VRF key will be used for block signing
+    if (vrfSignIsNotCorrect) {
+      throw new IllegalArgumentException(s"Genesis block timestamp is not signed his own forger box")
+    }
+  }
+
+  private def validateNonGenesisBlock(verifiedBlock: SidechainBlock, history: SidechainHistory): Unit = {
+    val parentBlockInfo: SidechainBlockInfo = history.storage.blockInfoById(verifiedBlock.parentId)
+    verifyTimestamp(verifiedBlock.timestamp, parentBlockInfo.timestamp, history)
+
+    val blockInfo = history.blockToBlockInfo(verifiedBlock)
+      .getOrElse(throw new IllegalArgumentException(s"Parent is missing for block ${verifiedBlock.id}")) //currently it is only reason if blockInfo is not calculated
+    val fullConsensusEpochInfo: FullConsensusEpochInfo = history.getFullConsensusEpochInfoForBlock(verifiedBlock.id, blockInfo)
+
+    verifyVrf(history, verifiedBlock, fullConsensusEpochInfo.nonceConsensusEpochInfo)
+    verifyForgerBox(verifiedBlock, fullConsensusEpochInfo.stakeConsensusEpochInfo)
+  }
+
+  private def verifyTimestamp(verifiedBlockTimestamp: Block.Timestamp, parentBlockTimestamp: Block.Timestamp, history: SidechainHistory): Unit = {
+    if (verifiedBlockTimestamp > Instant.now.getEpochSecond) throw new IllegalArgumentException("Block had been generated in the future")
+    if (verifiedBlockTimestamp < parentBlockTimestamp) throw new IllegalArgumentException("Block had been generated before parent block had been generated")
```

Comment:
verify also that slot number of current block is greater than the parent block

Reply:
done